google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/AddPerformanceMaxProductListingGroupTree.java

```diff
@@ -0,0 +1,556 @@
+              // .setParentListingGroupFilter("PARENT_FILTER_NAME")
+              //
+              // Unlike AddPerformanceMaxRetailCampaign, the type for the root node here must be
+              // SUBDIVISION because it will have child partitions under it.
+              .setType(ListingGroupFilterType.SUBDIVISION)
+              // Specifies that this is in the shopping vertical because it is a Performance Max
+              // campaign for retail.
+              .setVertical(ListingGroupFilterVertical.SHOPPING)
+              // Note the case_value is not set because it should be undefined for the root node.
+              .build();
+      AssetGroupListingGroupFilterOperation operation =
+          AssetGroupListingGroupFilterOperation.newBuilder().setCreate(listingGroupFilter).build();
+      return MutateOperation.newBuilder()
+          .setAssetGroupListingGroupFilterOperation(operation)
+          .build();
+    }
+    // [END add_performance_max_product_listing_group_tree_4]
+
+    // [START add_performance_max_product_listing_group_tree_5]
+    /**
+     * Creates a MutateOperation that creates an intermediate AssetGroupListingGroupFilter for the
+     * factory's AssetGroup.
+     *
+     * <p>Use this method if the filter will have child filters. Otherwise, use the {@link
+     * #createUnit(long, long, ListingGroupFilterDimension), createUnit} method.
+     *
+     * @param parent the ID of the parent AssetGroupListingGroupFilter.
+     * @param id the ID of AssetGroupListingGroupFilter that will be created.
+     * @param dimension the dimension to associate with the AssetGroupListingGroupFilter.
+     */
+    private MutateOperation createSubdivision(
+        long parent, long id, ListingGroupFilterDimension dimension) {
+      AssetGroupListingGroupFilter listingGroupFilter =
+          AssetGroupListingGroupFilter.newBuilder()
+              .setResourceName(
+                  ResourceNames.assetGroupListingGroupFilter(customerId, assetGroupId, id))
+              .setAssetGroup(ResourceNames.assetGroup(customerId, assetGroupId))
+              .setParentListingGroupFilter(
+                  ResourceNames.assetGroupListingGroupFilter(customerId, assetGroupId, parent))
+              // Uses the SUBDIVISION type to indicate that the AssetGroupListingGroupFilter
+              // will have children.
+              .setType(ListingGroupFilterType.SUBDIVISION)
+              // Specifies that this is in the shopping vertical because it is a Performance Max
+              // campaign for retail.
+              .setVertical(ListingGroupFilterVertical.SHOPPING)
+              .setCaseValue(dimension)
+              .build();
+      AssetGroupListingGroupFilterOperation filterOperation =
+          AssetGroupListingGroupFilterOperation.newBuilder().setCreate(listingGroupFilter).build();
+      return MutateOperation.newBuilder()
+          .setAssetGroupListingGroupFilterOperation(filterOperation)
+          .build();
```

Comment:
I personally find this style of builder construction more legible (always using newlines for different methods). I would encourage you to use the same style in other builder locations, but ultimately up to you.

Reply:
This is auto-formatted.